### PR TITLE
backfill testcoverage for node IP discovery

### PR DESCRIPTION
**Pull request by tylerschultz**
This PR adds tests for nodemanager.go functionality for what IPs are added to the Kubernetes Node object. Our intention is to later submit a PR for support for dualstack clusters. We want to make sure we don't cause any regressions from its current behavior.

**What this PR does / why we need it**:

We've added many test scenarios. Some of the scenarios we've added test coverage for have surprising results. We'd appreciate learning if there are any scenarios that reviewers think are edge cases with surprising results that we should not try to preserve. Even better would be to identify scenarios that we think have undesirable results that we can fix.

We'll start this PR, and then add comments to the scenarios we find interesting.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by tylerschultz** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

Please check my CLA again, Thanks!

File: pkg/cloudprovider/vsphere/nodemanager_test.go [hunk 1]
```diff
@@ -267,3 +268,825 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
...
+						Network: "external_net",
+						IpAddress: []string{
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "169.0.1.2"},
+				{Type: "ExternalIP", Address: "169.0.1.2"},
```
Comment (tylerschultz) [inline pkg/cloudprovider/vsphere/nodemanager_test.go:540]: This scenario is surprising to us, it seems the desired IPs are not selected, and instead default selection happens.

File: pkg/cloudprovider/vsphere/nodemanager_test.go [hunk 2]
```diff
@@ -267,3 +268,825 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
...
+							"172.15.108.10",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.1.22"},
+				{Type: "ExternalIP", Address: "10.10.1.22"},
+			},
```
Comment (tylerschultz) [inline pkg/cloudprovider/vsphere/nodemanager_test.go:799]: This seems unexpected. The desired IP exists on the second NIC.

File: pkg/cloudprovider/vsphere/nodemanager_test.go [hunk 3]
```diff
@@ -267,3 +268,825 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
...
+							"33.33.33.33",
+							"10.10.1.22",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "22.22.22.22"},
+				{Type: "ExternalIP", Address: "172.15.108.11"},
+			},
```
Comment (tylerschultz) [inline pkg/cloudprovider/vsphere/nodemanager_test.go:901]: This seems like undesireable behavior. The IPs that are returned
depend on ordering within the NICs or which NIC they are on.
There is no sensible precedence with the configuration.

File: pkg/cloudprovider/vsphere/nodemanager_test.go [hunk 4]
```diff
@@ -267,3 +268,825 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
...
+						IpAddress: []string{
+							"fe80:cccc::2",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "172.15.108.11"},
+				{Type: "ExternalIP", Address: "172.15.108.11"},
+			},
```
Comment (tylerschultz) [inline pkg/cloudprovider/vsphere/nodemanager_test.go:1025]: We believe this is a scenario where a prior (and reverted) dualstack PR caused an undesirable change in behavior.

File: pkg/cloudprovider/vsphere/nodemanager_test.go (hunk 1 above)
Comment (andrewsykim) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:540]: This seems like a bug to be honest :thinking:

Comment (andrewsykim) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:540]: Hooray for tests that catch bugs :)

File: pkg/cloudprovider/vsphere/nodemanager_test.go (hunk 3 above)
Comment (andrewsykim) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:901]: I think defining the precedence here is warranted.

**Review by andrewsykim — COMMENTED** [review]
Thanks so much for adding these tests, we really needed them for this code path

**Comment by andrewsykim** [comment]
/assign @lubronzhan @nicolehanjing 

PTAL as well!

File: pkg/cloudprovider/vsphere/nodemanager_test.go (hunk 1 above)
Comment (nicolehanjing) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:540]: yeah +1 - I believe this is a bug

File: pkg/cloudprovider/vsphere/nodemanager_test.go (hunk 3 above)
Comment (nicolehanjing) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:934]: I feel like there's a bug in the code - I think the ExternalIP is expected as we prioritize address masking over network name. However, for InternalIP, it looks like the network name has higher priority, which does not look right to me

File: pkg/cloudprovider/vsphere/nodemanager_test.go (hunk 2 above)
Comment (nicolehanjing) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:832]: 🤔  after checking the code I believe the expected IP should be "172.15.108.10"

Comment (tylerschultz) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:832]: @nicolehanjing I agree, this is incorrect behavior. I'd like to keep this PR as merely test coverage (noting this expectation is undesireable), and then follow up with a PR to refactor and correct these problems. The PR will highlight the bug fixes/changes in behavior.

File: pkg/cloudprovider/vsphere/nodemanager_test.go (hunk 3 above)
Comment (tylerschultz) [reply pkg/cloudprovider/vsphere/nodemanager_test.go:934]: I agree, this is incorrect behavior. I'd like to keep this PR as merely test coverage (noting this expectation is undesireable), and then follow up with a PR to refactor and correct these problems. The refactor PR diff will highlight the bug fixes/changes in behavior in the test.
